site/src/pages/WorkspacePage/useResourcesNav.ts

```diff
@@ -0,0 +1,44 @@
+import { WorkspaceResource } from "api/typesGenerated";
+import { useTab } from "hooks";
+import { useCallback, useEffect } from "react";
+
+export const resourceOptionId = (resource: WorkspaceResource) => {
+  return `${resource.type}_${resource.name}`;
+};
+
+export const useResourcesNav = (resources: WorkspaceResource[]) => {
+  const resourcesNav = useTab("resources", "");
+  const selectedResource = resources.find(
+    (r) => resourceOptionId(r) === resourcesNav.value,
+  );
+
+  useEffect(() => {
+    const hasResourcesWithAgents =
+      resources.length > 0 &&
+      resources[0].agents &&
+      resources[0].agents.length > 0;
+    if (!selectedResource && hasResourcesWithAgents) {
+      resourcesNav.set(resourceOptionId(resources[0]));
+    }
+  }, [resources, selectedResource, resourcesNav]);
```

Comment:
Since the only cue for synchronization seems to be when `selectedResource` changes, could this be redefined with `useEffectEvent` to make sure `useEffect` doesn't run too often?

```ts
const syncSelectionChange = useEffectEvent((previousResource) => {
  const hasResourcesWithAgents =
    resources.length > 0 &&
    resources[0].agents &&
    resources[0].agents.length > 0;
  if (!previousResource && hasResourcesWithAgents) {
    resourcesNav.set(resourceOptionId(resources[0]));
  }
});

useEffect(() => {
  syncSelectionChange(selectedResource);
}, [syncSelectionChange, selectedResource]);
```
Even if `resources` changes by array reference, if the selected resource is still in there, I wouldn't think that would be a cause for causing the effect to re-run. And because the `set` function doesn't have memoization, it would cause the effect to run on every render

Though I think there's one more edge case: what would happen if we had a resource selected, but then the resources array becomes completely empty? Would we need to set things back to an empty string?

Reply:
Overall, though, I think the effect logic is what's pushing me to think that a custom hook might be justified, mainly because it's relying on URL syncs via React Router. We can't really get away with derived values or inline state syncs because we'd have to talk to an outside system during the renders

I do think there's value in abstracting over those syncs and giving you a guarantee that things will "just work". I'm just wondering if it's worth testing the hook itself, or if it's better to treat it like an implementation detail, and test the component itself

Reply:
Yes, this would be the ideal scenario but rendering the workspace component and testing this behavior requires some server events mocking which I would like to avoid. I rather have a simple solution like this for now and only go for more complex ones if it is needed.

Reply:
> Though I think there's one more edge case: what would happen if we had a resource selected, but then the resources array becomes completely empty? Would we need to set things back to an empty string?

Hm... I think this would not happen because when resources are returned, we can expect at least one resource. I think Coder does not allow workspace creations without a resource but if it does, I think it is not a huge concern right now.

Reply:
> And because the set function doesn't have memoization, it would cause the effect to run on every render

I would not care too much about effects running on every render if it is not going to mess up with the state or cause a loop. I would rather optimize it to be ok running on every render one million times.